components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -0,0 +1,190 @@
+    }
+    textLabel.text = text;
+  } else {
+    _text = nil;
+    [self.textLabel removeFromSuperview];
+  }
+  [self.layout reloadData];
+  [self setNeedsLayout];
+}
+
+- (void)setImage:(UIImage *)image {
+  if (image) {
+    _image = image;
+    if (!self.iconImageViewContainer) {
+      CGRect iconImageContainerFrame =
+          CGRectMake(0, 0, kIconImageContainerSideLength, kIconImageContainerSideLength);
+      UIView *iconImageViewContainer = [[UIView alloc] initWithFrame:iconImageContainerFrame];
+      UIImageView *iconImageView = [[UIImageView alloc] initWithImage:image];
+      CGRect iconImageFrame = CGRectMake(0, 0, kIconImageSideLength, kIconImageSideLength);
+      iconImageView.frame = iconImageFrame;
+      iconImageView.contentMode = UIViewContentModeScaleAspectFit;
+      iconImageView.clipsToBounds = YES;
+      [iconImageViewContainer addSubview:iconImageView];
+      self.iconImageViewContainer = iconImageViewContainer;
+      self.iconImageView = iconImageView;
+      [self addSubview:iconImageViewContainer];
+    } else {
+      self.iconImageView.image = image;
+    }
+  } else {
+    _image = nil;
+    [self.iconImageViewContainer removeFromSuperview];
+    self.iconImageViewContainer = nil;
+  }
+  [self.layout reloadData];
+  [self setNeedsLayout];
+}
+
+- (void)setButtons:(NSArray<__kindof UIButton *> *)buttons {
+  for (UIButton *button in _buttons) {
+    [button removeFromSuperview];
+  }
+  for (UIButton *button in buttons) {
+    [self addSubview:button];
+  }
+  _buttons = buttons;
+  [self.layout reloadData];
+  [self setNeedsLayout];
+}
+
+#pragma mark - Banner Views
+
++ (MDCBannerView *)bannerWithText:(NSString *)text
+                            image:(UIImage *)image
+                          buttons:(NSArray<__kindof UIButton *> *)buttons {
+  MDCBannerView *mdcBannerView = [[MDCBannerView alloc] initWithFrame:CGRectZero];
+  mdcBannerView.text = text;
+  mdcBannerView.image = image;
+  mdcBannerView.buttons = buttons;
+  return mdcBannerView;
+}
+
+#pragma mark - UIView overrides
+
+- (CGSize)sizeThatFits:(CGSize)size {
+  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size];
+  self.layout.dataSource = self;
+  [self.layout reloadData];
+  return [self.layout frameSize];
+}
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
+
+  [self applyMDCBannerViewLayout:self.layout];
+}
+
+#pragma mark - MDCBannerViewLayoutDataSource
+
+- (UILabel *)textLabelForBannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout {
+  return self.textLabel;
+}
+
+- (UIView *)imageContainerForBannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout {
+  return self.iconImageViewContainer;
+}
+
+- (NSInteger)numberOfButtonsForBannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout {
+  return self.buttons.count;
+}
+
+- (UIButton *)bannerViewLayout:(MDCBannerViewLayout *)bannerViewLayout
+                 buttonAtIndex:(NSInteger)index {
+  return self.buttons[index];
+}
```

Comment:
I'm not sure we're gaining much from using a data source for the layout object. Consider exposing properties on the layout object instead and setting them directly.

Reply:
After re-evaluation, we are going to remove the datasource structure. Will address.

Reply:
Addressed in c26d9d14efafdeba2cd86334c4e4215592a41e83